Offline hash handling

GarethDaviesGW opened the pull request:
Makes a function which detects whether using python 2 or python 3 and decides whether to use native hash (unsafe in python 3) or hashlib module (utilises int.to_bytes which is unavailable in python2)

Solves issue #3302

Note that I haven't been able to test using bank2hdf, as I haven't been able to get that to work even with the original function! However this works with Tito's test script

titodalcanton commented:
As done in PyCBC Live, I recommend an explicit check for a hash collision using `assert` or `raise` as appropriate.

ahnitz commented on pycbc/waveform/bank.py:
```diff
@@ -178,6 +179,26 @@ def parse_approximant_arg(approximant_arg, warray):
     """
     return warray.parse_boolargs(boolargs_from_apprxstr(approximant_arg))[0]
 
+def list_to_hash(list_to_be_hashed):
```
To avoid confusion, I don't think you should call this 'list_XXX'. Hashing a list doesn't actually work. It really does have to be a tuple.

GarethDaviesGW commented:
I managed to get bank2hdf working - which caused an issue as the 64-byte integer being made wouldn't fit into a native numpy datatype. I've updated which hash algorithm is being used to make the hash.

Note that though a 8-byte (64 bit) hash _looks_ unsafe, it's actually slightly better than the in-built hash which uses 62 bits

I've changed the name as suggest by Alex, and added to raise an error as suggested by Tito

ahnitz commented:
@GarethDaviesGW I don't really see any problems with this patch. Have you checked that nothing crops up with a small workflow in python 3?

GarethDaviesGW commented:
I havent done a small-workflow test yet
I'll try to set one up

GarethDaviesGW commented:
Note that I haven't forgotten this, just taking a while to get it through the workflow on CIT

GarethDaviesGW commented:
I've had to add a couple of fixes for using python3 in multiifo_statmap_inj and cluster_coincs_multiifo, and one bugfix in pycbc_ifar_catalog for the case where there are no hierarchical removals

I've added these here just because they came up when running the test workflow for the hash handling change

GarethDaviesGW commented:
(note I am re-requesting a review for the other things I had to add)

ahnitz commented:
@GarethDaviesGW The last changes seem very straightforward, so I am happy to approve.
